presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/FuseHashFunctions.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.iterative.rule;
+
+import com.facebook.presto.sql.tree.Expression;
+import com.facebook.presto.sql.tree.ExpressionTreeRewriter;
+import com.facebook.presto.sql.tree.FunctionCall;
+import com.facebook.presto.sql.tree.QualifiedName;
+import com.google.common.collect.ImmutableList;
+
+import static com.google.common.collect.Iterables.getOnlyElement;
+
+public class FuseHashFunctions
+        extends ExpressionRewriteRuleSet
+{
+    public FuseHashFunctions()
+    {
+        super(createRewrite());
+    }
+
+    public static Expression rewrite(Expression expression)
+    {
+        return createRewrite().rewrite(expression, null);
+    }
+
+    private static ExpressionRewriter createRewrite()
+    {
+        return (expression, context) -> ExpressionTreeRewriter.rewriteWith(new com.facebook.presto.sql.tree.ExpressionRewriter<Void>()
+        {
+            @Override
+            public Expression rewriteFunctionCall(FunctionCall node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+            {
+                if (node.getName().equals(QualifiedName.of("from_big_endian_64"))) {
```

Comment:
make happy path not part of an if? e.g:
```
if (!node.getName().equals(QualifiedName.of("from_big_endian_64"))) {
  return node;
}

return rewriteFromBigEndian64(node, context, treeRewriter);
```

the same for other methods.  It's easier to read then nested iffs

Reply:
I agree with you in general, though I tried originally and it didn't seem better. I also structured it like this so we can add additional hash functions later.

It seems more straight forward to say, "if it's function X, rewrite X". Of course, this would be much more natural as pattern matching.

Here's what it looks like the other way:
```java
public Expression rewriteFunctionCall(FunctionCall node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
{
    if (!node.getName().equals(QualifiedName.of("from_big_endian_64"))) {
        return node;
    }
    return rewriteFromBigEndian64(node, context, treeRewriter);
}

private static Expression rewriteFromBigEndian64(FunctionCall node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
{
    Expression argument = treeRewriter.rewrite(getOnlyElement(node.getArguments()), context);
    if (!(argument instanceof FunctionCall)) {
        return node;
    }
    FunctionCall function = (FunctionCall) argument;
    if (!function.getName().equals(QualifiedName.of("xxhash64"))) {
        return node;
    }
    return rewriteXxhash64(function, context, treeRewriter);
}

private static Expression rewriteXxhash64(FunctionCall node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
{
    Expression argument = treeRewriter.rewrite(getOnlyElement(node.getArguments()), context);
    if (!(argument instanceof FunctionCall)) {
        return node;
    }
    FunctionCall function = (FunctionCall) argument;
    if (function.getName().equals(QualifiedName.of("to_big_endian_64"))) {
        argument = treeRewriter.rewrite(getOnlyElement(function.getArguments()), context);
    }
    return new FunctionCall(QualifiedName.of("$xxhash64"), ImmutableList.of(argument));
}

Reply:
TBH I like the latter code more (list negative conditions that cause early return and then have actual logic), but it's a matter of personal preference I guess.